### PR TITLE
[feature](nereids): add join rules base code

### PR DESCRIPTION
**Pull request by jackwener**
# Proposed changes

Issue Number: close #xxx

## Problem Summary:

Add base code about join reoreder of top-down

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (No Need)
3. Has document been added or modified: (No Need)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinCommutative.java [hunk 1]
```diff
@@ -15,28 +15,43 @@
...
+    private boolean justApplyInnerOuterCrossJoin = false;
+
+    private final SwapType swapType;
+
+    /**
+     * If param is true, just apply rule in inner/full-outer/cross join.
+     */
+    public JoinCommutative(boolean justApplyInnerOuterCrossJoin) {
+        this.justApplyInnerOuterCrossJoin = justApplyInnerOuterCrossJoin;
+        this.swapType = SwapType.ALL;
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinCommutative.java:39]: What's this used for ?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinExchange.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    /*
+     *        topJoin                      newTopJoin
+     *        /      \                      /      \
+     *   leftJoin  rightJoin   -->   newLeftJoin newRightJoin
+     *    /    \    /    \            /    \        /    \
+     *   A      B  C      D          A      C      B      D
+     */
+    @Override
+    public Rule<Plan> build() {
+        return innerLogicalJoin(innerLogicalJoin(), any()).then(topJoin -> {
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinExchange.java:40]: Is it correct?
Maybe it should be ```innerLogicalJoin(innerLogicalJoin(), innerLogicalJoin()) ```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinCommutative.java [hunk 3]
```diff
@@ -15,28 +15,43 @@
...
 import org.apache.doris.nereids.rules.RuleType;
-import org.apache.doris.nereids.trees.plans.JoinType;
+import org.apache.doris.nereids.rules.exploration.OneExplorationRuleFactory;
 import org.apache.doris.nereids.trees.plans.Plan;
 import org.apache.doris.nereids.trees.plans.logical.LogicalJoin;
 
 /**
  * rule factory for exchange inner join's children.
  */
 public class JoinCommutative extends OneExplorationRuleFactory {
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinCommutative.java:29]: Could you please add some unit test for those rule ?

Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinCommutative.java:29]: The current code is just an example. 
Can't run.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinCommutative.java (hunk 1 above)
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinCommutative.java:39]: limit swap space, we should change `JoinReorderContext ` according to it 
- left deep tree -> bottom join and disable l-asscom
- zig-zag ...
- busy...

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

**Pull request merged**